Python/Product/Uap/Debugger/PythonRemoteDebugEvents.cs

```diff
@@ -0,0 +1,118 @@
+        // remote Python debug server to have automatic attach work
+        internal const uint RemoteDebugStartExceptionCode = 0xEDCBA987;
+
+        public const string RemoteDebugExceptionId = "E7DD0845-FB1A-4A45-8192-44953C0ACC51";
+        public static readonly Guid RemoteDebugExceptionGuid = new Guid(RemoteDebugExceptionId);
+
+        public static PythonRemoteDebugEvents Instance {
+            get { return instance.Value; }
+        }
+
+        public Func<System.Threading.Tasks.Task> AttachRemoteProcessFunction { get; set; }
+
+        public string AttachRemoteDebugXml { get; set; }
+
+        public int Event(IDebugEngine2 pEngine, IDebugProcess2 pProcess, IDebugProgram2 pProgram, IDebugThread2 pThread, IDebugEvent2 pEvent, ref Guid riidEvent, uint dwAttrib) {
+            try {
+                if (riidEvent == typeof(IDebugProgramCreateEvent2).GUID) {
+                    Guid processId;
+
+                    // A program was created and attached
+                    if (pProcess != null) {
+                        if (VSConstants.S_OK == pProcess.GetProcessId(out processId)) {
+                            DkmProcess dkmProcess = DkmProcess.FindProcess(processId);
+
+                            if (dkmProcess != null) {
+                                var debugTrigger = DkmExceptionCodeTrigger.Create(DkmExceptionProcessingStage.Thrown, null, DkmExceptionCategory.Win32, RemoteDebugStartExceptionCode);
+
+                                // Try to add exception trigger for when a remote debugger server is started for Python
+                                dkmProcess.AddExceptionTrigger(RemoteDebugExceptionGuid, debugTrigger);
+                            }
+                        }
+                    }
+                }
+
+                return VSConstants.S_OK;
+            } finally {
+                if (pEngine != null && Marshal.IsComObject(pEngine)) {
+                    Marshal.ReleaseComObject(pEngine);
+                }
+
+                if (pProcess != null && Marshal.IsComObject(pProcess)) {
+                    Marshal.ReleaseComObject(pProcess);
+                }
+
+                if (pProgram != null && Marshal.IsComObject(pProgram)) {
+                    Marshal.ReleaseComObject(pProgram);
+                }
+
+                if (pThread != null && Marshal.IsComObject(pThread)) {
+                    Marshal.ReleaseComObject(pThread);
+                }
+
+                if (pEvent != null && Marshal.IsComObject(pEvent)) {
+                    Marshal.ReleaseComObject(pEvent);
+                }
+            }
+        }
+
+        void IDkmExceptionTriggerHitNotification.OnExceptionTriggerHit(DkmExceptionTriggerHit hit, DkmEventDescriptorS eventDescriptor) {
+            var remoteProcessTask = default(System.Threading.Tasks.Task);
+
+            using (var evt = new System.Threading.ManualResetEvent(false)) {
+
+                ThreadHelper.Generic.Invoke(() => {
```

Comment:
The invoked part has to run before the outer method returns, because at this point the debuggee is paused, and returning will unpause it. So this really needs a synchronous Invoke or something equivalent that will block the caller.